constraint-solver/src/constraint_system.rs

```diff
@@ -0,0 +1,141 @@
+    /// this should evaluate to 1 or -1.
+    pub multiplicity: V,
+}
+
+impl<V> BusInteraction<V> {
+    pub fn iter(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        Box::new(
+            [&self.bus_id, &self.multiplicity]
+                .into_iter()
+                .chain(self.payload.iter()),
+        )
+    }
+
+    pub fn iter_mut(&mut self) -> Box<dyn Iterator<Item = &mut V> + '_> {
+        Box::new(
+            [&mut self.bus_id, &mut self.multiplicity]
+                .into_iter()
+                .chain(self.payload.iter_mut()),
+        )
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord> BusInteraction<QuadraticSymbolicExpression<T, V>> {
+    /// Converts a bus interactions with fields represented by expressions
+    /// to a bus interaction with fields represented by range constraints.
+    pub fn to_range_constraints(
+        &self,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Option<BusInteraction<RangeConstraint<T>>> {
+        // TODO: Handle bus interactions with complex expressions.
+        Some(BusInteraction {
+            bus_id: expr_to_range_constraint(&self.bus_id, range_constraints)?,
+            payload: self
+                .payload
+                .iter()
+                .map(|expr| expr_to_range_constraint(expr, range_constraints))
+                .collect::<Option<Vec<_>>>()?,
+            multiplicity: expr_to_range_constraint(&self.multiplicity, range_constraints)?,
+        })
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord + Eq>
+    BusInteraction<QuadraticSymbolicExpression<T, V>>
+{
+    /// Attempts to execute the bus interaction, using the given
+    /// `BusInteractionHandler`. Returns a list of updates to be
+    /// executed by the caller.
+    pub fn solve(
+        &self,
+        bus_interaction_handler: &dyn BusInteractionHandler<T = T, V = V>,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Vec<Effect<T, V>> {
+        let Some(range_constraints) = self.to_range_constraints(range_constraints) else {
+            return vec![];
+        };
+        let range_constraints = bus_interaction_handler.handle_bus_interaction(range_constraints);
+        self.iter()
+            .zip(range_constraints.iter())
+            .filter_map(|(expr, rc)| {
+                if let Some(var) = expr.try_to_simple_unknown() {
+                    return Some(Effect::from_range_constraint_update(var, rc.clone()));
+                }
+                None
+            })
+            .collect()
+    }
+}
+
+/// A trait for handling bus interactions.
+pub trait BusInteractionHandler {
+    type T: FieldElement;
+    type V;
+
+    /// Handles a bus interaction, by transforming taking a bus interaction
+    /// (with the fields represented by range constraints) and returning
+    /// updated range constraints.
+    /// The range constraints are intersected with the previous ones by the
```

Comment:
Maybe add something like: The idea is that a certain combination of range constraints on elements can be further restricted given internal knowledge about the specific bus interaction.